### DISCUSSION TITLE
How can i modify config based on VGG16 ?

### DISCUSSION THREAD

**Discussion by YusukeImadegawa** [open-mmlab/mmclassification#1502] [Q&A] [answered]
Thank you for your help.
I fine-tuned vgg16 model to classify 10classes, and run 
① `!python tools/analysis_tools/get_flops.py configs/vgg/vgg16_b32x8_imagenet_building.py` 
② `!python tools/train.py configs/vgg/vgg16_b32x8_imagenet_building.py `
there are some Error : 

①get_flops.py
```
04/19 06:44:38 - mmengine - WARNING - Unsupported operator aten::max_pool2d encountered 5 time(s)
04/19 06:44:38 - mmengine - WARNING - The following submodules of the model were never called during the trace of the graph. They may be unused, or they were accessed by direct calls to .forward() or via other python methods. In the latter case they will have zeros for statistics, though their statistics will still contribute to their parent calling module.
data_preprocessor, head, head.fc, head.loss_module
```
②train.py
`RuntimeError: mat1 and mat2 shapes cannot be multiplied (32x1000 and 25088x10)`

there are a config i wrote, and visualization of structure below.
My image size is 224*224 rgb.
# Config
```
%%writefile configs/vgg/vgg16_b32x8_imagenet_building.py
_base_ = [
    '../_base_/models/vgg16.py', 
    '../_base_/datasets/imagenet_bs256_davit_224.py', 
    '../_base_/schedules/imagenet_bs256_epochstep.py',
    '../_base_/default_runtime.py'
]

model = dict(
    type='ImageClassifier', 
    backbone=dict(
        type='VGG',
        depth=16,
        init_cfg=dict( 
            type='Pretrained',
            checkpoint='https://download.openmmlab.com/mmclassification/v0/vgg/vgg16_batch256_imagenet_20210208-db26f1a5.pth',
            prefix='backbone',
        ),
        frozen_stages=2, 
    ),
    head=dict( 
        type='LinearClsHead', 
        num_classes=10, 
        in_channels=512*7*7, 
        topk=(1, ), 
        cal_acc=True), 
)

data_preprocessor = dict(
    mean=[124.508, 116.050, 106.438], 
    std=[58.577, 57.310, 57.437], 
    to_rgb=True,
)

train_pipeline = [
    dict(type='LoadImageFromFile'), 
    dict(type='RandomResizedCrop', size=224, backend='pillow'),
    dict(type='RandomFlip', flip_prob=0.5, direction='horizontal'), 
    dict(type='ImageToTensor', keys=['img']), 
    dict(type='ToTensor', keys=['gt_label']), 
    dict(type='Collect', keys=['img', 'gt_label']) 
]

val_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=(256, -1), backend='pillow'), 
    dict(type='CenterCrop', crop_size=224),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]

test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=(256, -1), backend='pillow'),
    dict(type='CenterCrop', crop_size=224),
    dict(type='ImageToTensor', keys=['img']),
    dict(type='Collect', keys=['img'])
]

data_root = '/content/drive/MyDrive/MMClassification/my_dataset'
train_dataloader = dict(
    batch_size=32,
    num_workers=2, 
    dataset=dict(
        type='CustomDataset',
        data_root=data_root,
        data_prefix='train/',
        ann_file='meta/trainQ1.txt',
        classes=[0,1,2,3,4,5,6,7,8,9],
    ))

val_dataloader = dict(
    batch_size=32,
    num_workers=2, 
    dataset=dict(
        type='CustomDataset',
        data_root=data_root,
        data_prefix='val/',
        ann_file='meta/valQ1.txt',
        classes=[0,1,2,3,4,5,6,7,8,9],
    ))

test_dataloader = dict(
    batch_size=32,
    num_workers=2, 
    dataset=dict(
        type='CustomDataset',
        data_root=data_root,
        data_prefix='test/',
        ann_file='meta/testQ1.txt',
        classes=[0,1,2,3,4,5,6,7,8,9],
    ))

val_evaluator = dict(type='Accuracy', topk=(1, ))

# optimizer hyper-parameters
optim_wrapper = dict(
    optimizer=dict(type='SGD', lr=0.005, momentum=0.9, weight_decay=0.0001))
# learning policy
param_scheduler = dict(type='StepLR', step_size=6, gamma=0.1)

# runtime settings
train_cfg = dict(by_epoch=True, max_epochs=100, val_interval=1) 
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=10),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=1, max_keep_ckpts=3),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='VisualizationHook', enable=False))

randomness = dict(seed=0, diff_rank_seed=True)
```
# Visualization of model surautre
+---------------------------+----------------------+------------+--------------+
| module                    | #parameters or shape | #flops     | #activations |
+---------------------------+----------------------+------------+--------------+
| model                     | 0.138G               | 15.47G     | 13.557M      |
|  backbone                 |  0.138G              |  15.47G    |  13.557M     |
|   backbone.features       |   14.715M            |   15.347G  |   13.548M    |
|    backbone.features.0.c… |    1.792K            |    86.704M |    3.211M    |
|    backbone.features.1.c… |    36.928K           |    1.85G   |    3.211M    |
|    backbone.features.3.c… |    73.856K           |    0.925G  |    1.606M    |
|    backbone.features.4.c… |    0.148M            |    1.85G   |    1.606M    |
|    backbone.features.6.c… |    0.295M            |    0.925G  |    0.803M    |
|    backbone.features.7.c… |    0.59M             |    1.85G   |    0.803M    |
|    backbone.features.8.c… |    0.59M             |    1.85G   |    0.803M    |
|    backbone.features.10.… |    1.18M             |    0.925G  |    0.401M    |
|    backbone.features.11.… |    2.36M             |    1.85G   |    0.401M    |
|    backbone.features.12.… |    2.36M             |    1.85G   |    0.401M    |
|    backbone.features.14.… |    2.36M             |    0.462G  |    0.1M      |
|    backbone.features.15.… |    2.36M             |    0.462G  |    0.1M      |
|    backbone.features.16.… |    2.36M             |    0.462G  |    0.1M      |
|   backbone.classifier     |   0.124G             |   0.124G   |   9.192K     |
|    backbone.classifier.0  |    0.103G            |    0.103G  |    4.096K    |
|    backbone.classifier.3  |    16.781M           |    16.777M |    4.096K    |
|    backbone.classifier.6  |    4.097M            |    4.096M  |    1K        |
|  head.fc                  |  5.13K               |            |              |
|   head.fc.weight          |   (10, 512)          |            |              |
|   head.fc.bias            |   (10,)              |            |              |
+---------------------------+----------------------+------------+--------------+

==============================
Input shape: (3, 224, 224)
Flops: 15.47G
Params: 0.138G
Activation: 13.557M

**Comment by Ezra-Yu** [comment]
Use the following model config:

```
# model settings
model = dict(
    type='ImageClassifier',
    backbone=dict(
        type='VGG',
        depth=16,
        num_classes=10,
        init_cfg=dict( 
            type='Pretrained',
            checkpoint='https://download.openmmlab.com/mmclassification/v0/vgg/vgg16_batch256_imagenet_20210208-db26f1a5.pth',
            prefix='backbone',
        ),
        frozen_stages=2, 
    ),
    neck=None,
    head=dict(
        type='ClsHead',
        loss=dict(type='CrossEntropyLoss', loss_weight=1.0),
        topk=(1, 5),
    ))
```

**Reply by YusukeImadegawa** [reply]
Thank you for your help.
I modified my code to your part, but there is a error below. 
Could you give me some tips?
```
04/20 03:07:50 - mmengine - WARNING - The model and loaded state dict do not match exactly
size mismatch for classifier.6.weight: copying a param with shape torch.Size([1000, 4096]) from checkpoint, the shape in current model is torch.Size([10, 4096]).
size mismatch for classifier.6.bias: copying a param with shape torch.Size([1000]) from checkpoint, the shape in current model is torch.Size([10]).
```

**Reply by Ezra-Yu** [reply]
Your training is fine, this is just a warning, the initial pre-trained model weight output has 1000 categories, and your current task only has 10 categories, so this warning is reported.

**Reply by YusukeImadegawa** [reply, marked as answer]
Thank you enough, I'm relieved!
